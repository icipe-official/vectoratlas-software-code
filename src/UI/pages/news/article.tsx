import { Container } from '@mui/material';
import React from 'react';
import NewsDetails from '../../components/news/newsDetails';

const NewsArticlePage = (): JSX.Element => {
  return (
    <>
      <div>
        <main>
          <Container
            maxWidth={false}
            sx={{
              padding: '10px',
              maxWidth: '75%',
            }}
          >
            <NewsDetails />
          </Container>
        </main>
      </div>
    </>
  );
};

export default NewsArticlePage;
