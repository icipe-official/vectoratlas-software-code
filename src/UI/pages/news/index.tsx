import { Container } from '@mui/material';
import React from 'react';
import NewsList from '../../components/news/newsList';

const NewsPage = (): JSX.Element => {
  return (
    <>
      <div>
        <main>
          <Container
            maxWidth={false}
            sx={{
              padding: '10px',
              maxWidth: '75%',
            }}
          >
            <NewsList />
          </Container>
        </main>
      </div>
    </>
  );
};

export default NewsPage;
