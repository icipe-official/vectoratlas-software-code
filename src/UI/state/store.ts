import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import rootReducer from './rootReducer';

export function makeStore() {
  return configureStore({
    reducer: rootReducer,
  });
}

const store = makeStore();

export type AppState = ReturnType<typeof store.getState>;
export const rootInitialState = store.getState() as AppState;

export type AppDispatch = typeof store.dispatch;

export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  AppState,
  unknown,
  Action<string>
>;

export default store;
