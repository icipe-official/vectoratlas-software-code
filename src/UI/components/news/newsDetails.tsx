import { Button, Grid, Typography } from '@mui/material';
import CircularProgress from '@mui/material/CircularProgress';
import { useRouter } from 'next/router';
import React, { useEffect } from 'react';
import ReactMarkdown from 'react-markdown';
import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { getNews } from '../../state/news/actions/news.action';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';

export const NewsDetails = () => {
  const dispatch = useAppDispatch();
  const newsItem = useAppSelector((s) => s.news.currentNewsForEditing);
  const loadingNews = useAppSelector((s) => s.news.loading);
  const isEditor = useAppSelector((state) =>
    state.auth.roles.includes('editor')
  );

  const router = useRouter();
  const id = router.query.id as string | undefined;

  useEffect(() => {
    if (id) {
      dispatch(getNews(id));
    }
  }, [id, dispatch]);

  if (loadingNews || !newsItem) {
    return (
      <div style={{ width: '100%', display: 'flex', justifyContent: 'center' }}>
        <CircularProgress />
      </div>
    );
  }

  return (
    <div
      style={{
        paddingLeft: 4,
        paddingRight: 4,
        paddingTop: 1,
        paddingBottom: 10,
      }}
    >
      <div>
        <Button onClick={() => router.push('/news')}>
          <ArrowBackIcon />
          <Typography fontSize={'medium'}>Back to news list</Typography>
        </Button>
      </div>

      <div style={{ display: 'flex' }}>
        <div style={{ flexGrow: 1 }}>
          <ReactMarkdown
            components={{
              a: ({ node, ...props }) => (
                <a style={{ color: 'blue' }} {...props} />
              ),
              h2: ({ node, ...props }) => (
                <h2 style={{ margin: 0 }} {...props} />
              ),
            }}
          >
            {'## ' + newsItem.title}
          </ReactMarkdown>
        </div>
        {isEditor && (
          <Button
            variant="contained"
            style={{ whiteSpace: 'nowrap', height: '100%' }}
            onClick={() => router.push('/news/edit?id=' + newsItem.id)}
          >
            Edit item
          </Button>
        )}
      </div>
      <Grid container spacing={3}>
        <Grid item xs={12} lg={7} key="content">
          <ReactMarkdown
            components={{
              a: ({ node, ...props }) => (
                <a style={{ color: 'blue' }} {...props} />
              ),
              p: ({ node, ...props }) => (
                <p
                  style={{
                    marginTop: 15,
                    marginBottom: 0,
                    textAlign: 'justify',
                  }}
                  {...props}
                />
              ),
            }}
          >
            {newsItem.summary}
          </ReactMarkdown>
        </Grid>
        <Grid item xs={12} lg={5}>
          <picture>
            <img
              src={newsItem.image}
              style={{ width: '100%', paddingTop: '20px' }}
            />
          </picture>
        </Grid>
      </Grid>

      <ReactMarkdown
        components={{
          a: ({ node, ...props }) => <a style={{ color: 'blue' }} {...props} />,
        }}
      >
        {newsItem.article}
      </ReactMarkdown>
    </div>
  );
};

export default NewsDetails;
