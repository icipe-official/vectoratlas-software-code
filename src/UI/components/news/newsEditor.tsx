import { useRouter } from 'next/router';
import React, { useCallback, useEffect, useState } from 'react';
import { Button, TextField, Typography } from '@mui/material';
import { TextEditor } from '../shared/textEditor/RichTextEditor';
import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { getNews, upsertNews } from '../../state/news/actions/news.action';
import { News } from '../../state/state.types';
import { toast } from 'react-toastify';
import UploadIcon from '@mui/icons-material/Upload';
import CircularProgress from '@mui/material/CircularProgress';
import { toBase64 } from '../shared/imageTools';

const UPLOAD_LIMIT_IN_KB = 512;

const NewsEditor = () => {
  const [article, setArticle] = useState('');
  const [initialArticle, setInitialArticle] = useState('');
  const [summary, setSummary] = useState('');
  const [initialSummary, setInitialSummary] = useState('');
  const [image, setImage] = useState('');
  const [title, setTitle] = useState('');
  const dispatch = useAppDispatch();
  const currentNewsItem = useAppSelector((s) => s.news.currentNewsForEditing);
  const loadingNews = useAppSelector((s) => s.news.loading);

  const router = useRouter();
  const id = router.query.id as string | undefined;

  const saveNewsItem = useCallback(() => {
    const news: News = {
      id,
      title: title,
      summary: summary,
      article: article,
      image: image,
    };
    dispatch(upsertNews(news));
  }, [dispatch, id, article, title, summary, image]);

  const handleImageUpload = async (e: React.ChangeEvent<HTMLInputElement>) => {
    // only allow images below 512 KB
    if (e.target.files && e.target.files[0].size < UPLOAD_LIMIT_IN_KB * 1024) {
      const image = await toBase64(e.target.files[0]);
      setImage(image);
    } else {
      toast.error('Uploaded files must be less than 512 KB.', {
        autoClose: 5000,
      });
    }
  };

  useEffect(() => {
    if (id) {
      dispatch(getNews(id));
    }
  }, [dispatch, id]);

  useEffect(() => {
    if (currentNewsItem) {
      setTitle(currentNewsItem.title);
      setSummary(currentNewsItem.summary);
      setInitialSummary(currentNewsItem.summary);
      setArticle(currentNewsItem.article);
      setInitialArticle(currentNewsItem.article);
      setImage(currentNewsItem.image);
    }
  }, [currentNewsItem]);

  const titleValid = title !== '';
  const summaryValid = summary !== '';

  return (
    <div>
      <Typography variant="h4" sx={{ mt: 2, mb: 1 }}>
        {id ? 'Edit' : 'Create'} news item
      </Typography>
      {loadingNews ? (
        <div
          style={{ width: '100%', display: 'flex', justifyContent: 'center' }}
        >
          <CircularProgress />
        </div>
      ) : null}
      <Typography color="primary" variant="h5" sx={{ mt: 2, mb: 1 }}>
        Title
      </Typography>
      <TextField
        disabled={loadingNews}
        variant="outlined"
        sx={{ width: '100%' }}
        value={title}
        onChange={(e) => setTitle(e.target.value)}
        error={!titleValid}
        helperText={!titleValid ? 'Title cannot be empty' : ''}
      />
      <Typography color="primary" variant="h5" sx={{ mt: 2, mb: 1 }}>
        Summary
      </Typography>
      {!loadingNews ? (
        <TextEditor
          error={!summaryValid}
          helperText={!summaryValid ? 'Summary cannot be empty' : ''}
          description={summary}
          setDescription={setSummary}
          initialDescription={initialSummary}
        />
      ) : (
        <div style={{ height: 250 }} />
      )}
      <Typography color="primary" variant="h5" sx={{ mt: 2, mb: 1 }}>
        Article
      </Typography>
      {!loadingNews ? (
        <TextEditor
          description={article}
          setDescription={setArticle}
          initialDescription={initialArticle}
        />
      ) : (
        <div style={{ height: 250 }} />
      )}
      <Typography color="primary" variant="h5" sx={{ mt: 2, mb: 1 }}>
        Image
      </Typography>
      <div
        style={{
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
        }}
      >
        <Button
          disabled={loadingNews}
          variant="contained"
          component="label"
          style={{ width: '50%', minWidth: '250px' }}
        >
          <UploadIcon />
          Upload Image File
          <input
            data-testid="image-upload-input"
            type="file"
            hidden
            accept="image/*"
            onChange={handleImageUpload}
          />
        </Button>
        <Typography>
          Images must be smaller than {UPLOAD_LIMIT_IN_KB} KB.
        </Typography>
        {image ? (
          <picture>
            <img style={{ width: '30vw' }} src={image} alt="Article image" />
          </picture>
        ) : null}
      </div>
      <div
        style={{ display: 'flex', justifyContent: 'flex-end', marginTop: 15 }}
      >
        <Button
          variant="contained"
          disabled={loadingNews || !titleValid || !summaryValid}
          onClick={saveNewsItem}
          sx={{ m: 0, minWidth: 150 }}
        >
          {id ? 'Update' : 'Create'}
        </Button>
      </div>
    </div>
  );
};

export default NewsEditor;
