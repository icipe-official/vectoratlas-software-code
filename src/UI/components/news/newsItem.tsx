import React, { useState } from 'react';
import { News } from '../../state/state.types';
import {
  Button,
  CardContent,
  CardMedia,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Grid,
  IconButton,
  Paper,
  Typography,
} from '@mui/material';
import ReactMarkdown from 'react-markdown';
import EditIcon from '@mui/icons-material/Edit';
import { useRouter } from 'next/router';
import { useDispatch } from 'react-redux';
import { AppDispatch } from '../../state/store';
import { deleteNews } from '../../state/news/actions/news.action';

export const NewsItem = ({
  item,
  isEditor,
  hideMoreDetailsButton,
}: {
  item: News;
  isEditor: boolean;
  hideMoreDetailsButton?: boolean;
}) => {
  const router = useRouter();
  const dispatch = useDispatch<AppDispatch>();

  const handleEditClick = () => {
    router.push('/news/edit?id=' + item.id);
  };

  const handleMoreDetailsClick = () => {
    router.push('/news/article?id=' + item.id);
  };

  const [openDialog, setOpenDialog] = useState(false); // State to control dialog visibility
  const [selectedNewsId, setSelectedNewsId] = useState<string | null>(null); // State for selected species ID

  const handleDelete = (newsId: string) => {
    setSelectedNewsId(newsId); // Set selected species ID
    setOpenDialog(true); // Open the confirmation dialog
  };

  const handleConfirmDelete = () => {
    if (selectedNewsId) {
      dispatch(deleteNews(selectedNewsId)); // Dispatch the delete action
    }
    setOpenDialog(false); // Close the dialog after confirming delete
  };

  const handleCloseDialog = () => {
    setOpenDialog(false); // Close the dialog without deleting
  };

  return (
    <Paper sx={{ margin: 0, marginLeft: '2px' }}>
      <Grid
        container
        spacing={0}
        sx={{ justifyContent: 'center' }}
        className="BannerGrid"
      >
        <Grid item xs={12} md={9} key="content">
          <CardContent className="Content">
            <ReactMarkdown
              components={{
                a: ({ node, ...props }) => (
                  <a style={{ color: 'blue' }} {...props} />
                ),
                h2: ({ node, ...props }) => (
                  <h2 style={{ margin: 0 }} {...props} />
                ),
              }}
            >
              {'## ' + item.title}
            </ReactMarkdown>

            <ReactMarkdown
              components={{
                a: ({ node, ...props }) => (
                  <a style={{ color: 'blue' }} {...props} />
                ),
                p: ({ node, ...props }) => (
                  <p
                    style={{
                      marginTop: 15,
                      marginBottom: 0,
                      textAlign: 'justify',
                    }}
                    {...props}
                  />
                ),
              }}
            >
              {item.summary}
            </ReactMarkdown>
          </CardContent>
          <Grid sx={{ justifyContent: 'end', width: '100%', display: 'flex' }}>
            {!hideMoreDetailsButton && (
              <Button
                variant="outlined"
                onClick={handleMoreDetailsClick}
                className="ViewButton"
              >
                More details
              </Button>
            )}
            {isEditor && (
              <Button
                variant="contained"
                onClick={handleEditClick}
                className="EditButton"
              >
                Edit item
              </Button>
            )}
            {isEditor && (
              <Button
                sx={{ backgroundColor: 'red' }}
                variant="contained"
                onClick={() => handleDelete(item.id as string)}
                className="EditButton"
              >
                Delete item
              </Button>
            )}
          </Grid>
        </Grid>
        <Grid
          item
          xs={12}
          md={3}
          key={item.title}
          sx={{ height: '50vh', maxHeight: '40vh' }}
        >
          <CardMedia
            className="Media"
            image={item.image}
            sx={{ height: '100%', overflow: 'hidden' }}
          ></CardMedia>
        </Grid>
        {/* Dialog Box */}
        <Dialog
          open={openDialog}
          onClose={handleCloseDialog}
          aria-labelledby="alert-dialog-title"
          aria-describedby="alert-dialog-description"
        >
          <DialogTitle id="alert-dialog-title">{'Confirm Delete'}</DialogTitle>
          <DialogContent>
            <DialogContentText id="alert-dialog-description">
              Are you sure you want to delete this News? This action cannot be
              undone.
            </DialogContentText>
          </DialogContent>
          <DialogActions>
            <Button onClick={handleCloseDialog} color="primary">
              Cancel
            </Button>
            <Button onClick={handleConfirmDelete} color="secondary" autoFocus>
              Delete
            </Button>
          </DialogActions>
        </Dialog>
      </Grid>
    </Paper>
  );
};
