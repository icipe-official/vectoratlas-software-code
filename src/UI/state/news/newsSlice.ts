import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { News } from '../state.types';

export interface NewsState {
  currentNewsForEditing: News | null;
  loading: boolean;
  news: News[];
  topNews: News[];
}

export const initialState: () => NewsState = () => ({
  currentNewsForEditing: null,
  loading: false,
  news: [],
  topNews: [],
});

export const newsSlice = createSlice({
  name: 'news',
  initialState: initialState(),
  reducers: {
    setCurrentNewsForEditing(state, action) {
      state.currentNewsForEditing = action.payload;
    },
    newsLoading(state, action) {
      state.loading = action.payload;
    },
    setNewsItems(state, action) {
      state.news = action.payload;
    },
    setTopNewsItems(state, action) {
      state.topNews = action.payload;
    },
    clearCurrentNewsItem (state,action: PayloadAction<void>) {
      state.currentNewsForEditing = null;
    }
  },
  extraReducers: () => {},
});

export const {
  setCurrentNewsForEditing,
  newsLoading,
  setNewsItems,
  setTopNewsItems,
  clearCurrentNewsItem,
} = newsSlice.actions;

export default newsSlice.reducer;
