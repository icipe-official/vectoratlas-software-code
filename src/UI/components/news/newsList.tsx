import { Button, Paper, Typography } from '@mui/material';
import CircularProgress from '@mui/material/CircularProgress';
import React, { useEffect } from 'react';
import { useAppDispatch, useAppSelector } from '../../state/hooks';
import { getAllNewsItems } from '../../state/news/actions/news.action';
import { NewsItem } from './newsItem';

import router from 'next/router';
import { clearCurrentNewsItem } from '../../state/news/newsSlice';

export const NewsList = () => {
  const dispatch = useAppDispatch();
  
  const newsItems = useAppSelector((s) => s.news.news);
  const loadingNews = useAppSelector((s) => s.news.loading);
  const isEditor = useAppSelector((state) =>
    state.auth.roles.includes('editor')
  );

  useEffect(() => {
    dispatch(getAllNewsItems());
  }, [dispatch]);

  if (loadingNews) {
    return (
      <div style={{ width: '100%', display: 'flex', justifyContent: 'center' }}>
        <CircularProgress />
      </div>
    );
  }

  const createNewArticle = () => {
    dispatch(clearCurrentNewsItem())
    router.push('/news/edit');
  };

  return (
    <div>
      <div style={{ display: 'flex', alignItems: 'center' }}>
        <Typography
          color="primary"
          variant="h4"
          sx={{ mt: 2, mb: 1 }}
          style={{ flexGrow: 1 }}
        >
          News
        </Typography>
        {isEditor ? (
          <Button
            variant="contained"
            style={{ height: '50%' }}
            onClick={createNewArticle}
          >
            Create new article
          </Button>
        ) : null}
      </div>

      {newsItems.map((n) => (
        <div
          key={n.id}
          style={{
            paddingLeft: 4,
            paddingRight: 4,
            paddingTop: 1,
            paddingBottom: 10,
          }}
        >
          <NewsItem isEditor={isEditor} item={n} />
        </div>
      ))}
    </div>
  );
};

export default NewsList;
