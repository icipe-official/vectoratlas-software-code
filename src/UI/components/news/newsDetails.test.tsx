import React from 'react';
import { fireEvent, render } from '../../test_config/render';
import NewsDetails from './newsDetails';
import { AppState } from '../../state/store';
import { initialState } from '../../state/news/newsSlice';
import { useRouter } from 'next/router';

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    query: {},
  }),
}));

jest.mock(
  '@mui/material/CircularProgress',
  () =>
    function CircularProgressMock() {
      return <div>Circular progress mock</div>;
    }
);

jest.mock(
  'react-markdown',
  () =>
    function MockReactMarkdown(props: any) {
      return <div>React markdown mock {JSON.stringify(props)}</div>;
    }
);

jest.mock('../news/newsItem', () => ({
  NewsItem: function NewsItemMock(props) {
    return <div> News item mock {JSON.stringify(props)}</div>;
  },
}));

jest.mock('../../state/news/actions/news.action', () => ({
  getNews: function getNewsMock(id) {
    return { type: 'getNewsMock', payload: id };
  },
}));

describe('NewsDetails component', () => {
  let state: Partial<AppState>;

  beforeEach(() => {
    jest.clearAllMocks();

    state = {
      news: initialState(),
    };

    state.news.currentNewsForEditing = {
      id: '123-456',
      title: 'test news item',
      summary: 'test summary',
      image: 'image-base64-abc-123',
      article: 'test article',
    };
  });

  it('shows loading spinning when loading', () => {
    state.news.loading = true;

    const { wrapper } = render(<NewsDetails />, state);

    expect(wrapper.getByText('Circular progress mock')).toBeInTheDocument();
  });

  it('renders the news details correctly', () => {
    const { wrapper } = render(<NewsDetails />, state);

    expect(wrapper.container).toMatchSnapshot();
  });

  it('loads the details when first loading if there is an id', () => {
    (useRouter as jest.Mock).mockReturnValue({
      query: {
        id: '123-456',
      },
    });
    const { store } = render(<NewsDetails />, state);

    const actions = store.getActions();
    expect(actions[0]).toEqual({ type: 'getNewsMock', payload: '123-456' });
  });

  it('shows edit button if editor', () => {
    state.auth = {
      roles: ['editor'],
    };

    const { wrapper } = render(<NewsDetails />, state);

    expect(wrapper.getByText('Edit item')).toBeInTheDocument();
  });
});
