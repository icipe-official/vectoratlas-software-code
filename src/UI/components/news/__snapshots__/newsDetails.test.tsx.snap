// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsDetails component renders the news details correctly 1`] = `
<div>
  <div
    style="padding: 1px 4px 10px 4px;"
  >
    <div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          />
        </svg>
        <p
          class="MuiTypography-root MuiTypography-body1 css-18a2bkc-MuiTypography-root"
        >
          Back to news list
        </p>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      style="display: flex;"
    >
      <div
        style="flex-grow: 1;"
      >
        <div>
          React markdown mock 
          {"components":{},"children":"## test news item"}
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-7 css-117e5y7-MuiGrid-root"
      >
        <div>
          React markdown mock 
          {"components":{},"children":"test summary"}
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-lg-5 css-14tayut-MuiGrid-root"
      >
        <picture>
          <img
            alt="News article image"
            src="image-base64-abc-123"
            style="width: 100%; padding-top: 20px;"
          />
        </picture>
      </div>
    </div>
    <div>
      React markdown mock 
      {"components":{},"children":"test article"}
    </div>
  </div>
</div>
`;
